proposals/NNNN-sdl-protocol-security-specification.md

```diff
@@ -0,0 +1,394 @@
+    <td>0x02</td><td>2</td>
+    <td>Unknown Query ID</td>
+  </tr>
+  <tr>
+    <td>ERROR_NOT_SUPPORTED</td>
+    <td>0x03</td><td>3</td>
+    <td>SDL does not support encryption</td>
+  </tr>
+  <tr>
+    <td>ERROR_SERVICE_ALREADY_PROTECTED</td>
+    <td>0x04</td><td>4</td>
+    <td>got request to protect a service that was protected before</td>
+  </tr>
+  <tr>
+    <td>ERROR_SERVICE_NOT_PROTECTED</td>
+    <td>0x05</td><td>5</td>
+    <td>got handshake or encrypted data for not protected service</td>
+  </tr>
+  <tr>
+    <td>ERROR_DECRYPTION_FAILED</td>
+    <td>0x06</td><td>6</td>
+    <td>decryption failed</td>
+  </tr>
+  <tr>
+    <td>ERROR_ENCRYPTION_FAILED</td>
+    <td>0x07</td><td>7</td>
+    <td>encryption failed</td>
+  </tr>
+  <tr>
+    <td>ERROR_SSL_INVALID_DATA</td>
+    <td>0x08</td><td>8</td>
+    <td>SSL invalid data</td>
+  </tr>
+  <tr>
+    <td>ERROR_HANDSHAKE_FAILED</td>
+    <td>0x09</td><td>9</td>
+    <td>In case of all other handshake errors</td>
+  </tr>
+  <tr>
+    <td>INVALID_CERT</td>
+    <td>0x0A</td><td>10</td>
+    <td>handshake failed because certificate is invalid</td>
+  </tr>
+  <tr>
+    <td>EXPIRED_CERT</td>
+    <td>0x0B</td><td>11</td>
+    <td>handshake failed because certificate is expired</td>
+  </tr>
+  <tr>
+    <td>ERROR_INTERNAL</td>
+    <td>0xFF</td><td>255</td>
+    <td>internal error</td>
+  </tr>
+  <tr>
+    <td>ERROR_UNKNOWN_INTERNAL_ERROR</td>
+    <td>0xFE</td><td>254</td>
+    <td>Error value for testing</td>
+  </tr>
+</table>
+
+#### 4.7.5.4 Actions in case of an error
+
+In case of an error, the system and the application should reset the active SSL connection of the current transport connection. This impacts already established secured service sessions as all of them will be closed. The application will need to restart all services which require protection.
+
+## Potential downsides
+
+Identifying potential downsides are difficult as this proposal adds specifications and documentation to existing implementation. The downside of accepting this proposal is the effort that's required to review and audit the implementations of every SDL platform.
+
+## Impact on existing code
+
+SDL Core impact is very low as most of the specification is reverse engineered from it. Still following items impact SDL Core:
+1. A further review is required to improve error handling.
+2. Error codes around handshake failed or invalid/expired cert are new and should be used by SDL Core if these errors occur.
+3. A known issue should be resolved in that SDL Core doesn't respond with NAK if the application sends an error frame.
+
+Compared to Core, more effort is required on the SDL libraries. Their implementation is incomplete and contradicting to SDL Core and the protocol. For instance the sdl_ios library uses the RPC Payload header instead of a Query Header to send TLS handshake data. The structure of these headers don't align in the first byte (RPC Payload 4 bit, Query Header 8 bit).
```

Comment:
```suggestion
Compared to Core, more effort is required on the SDL app libraries. Their implementation is incomplete and contradicting to SDL Core and the protocol. For instance the sdl_ios library uses the RPC Payload header instead of a Query Header to send TLS handshake data. The structure of these headers don't align in the first byte (RPC Payload 4 bit, Query Header 8 bit).
```